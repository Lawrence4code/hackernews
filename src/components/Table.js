import React, { Component } from 'react';
import classNames from 'classnames';
import { sortBy } from 'lodash';

// Local imports
import Button from './Button';
import sortArrows from '../assets/icons/select-arrows.svg';

// sorting function
const Sort = ({ sortKey, activeSortKey, onSort, children }) => {
  const sortClass = classNames('button-inline', {
    'button-active': sortKey === activeSortKey
  });

  return (
    <Button onClick={() => onSort(sortKey)} className={sortClass}>
      {children}
    </Button>
  );
};

// object of function for sorting
const SORTS = {
  NONE: list => list,
  TITLE: list => sortBy(list, 'title'),
  AUTHOR: list => sortBy(list, 'author'),
  COMMENTS: list => sortBy(list, 'num_comments').reverse(),
  POINTS: list => sortBy(list, 'points').reverse()
};

class Table extends Component {
  constructor(props) {
    super(props);
    // table component state
    this.state = {
      sortKey: 'NONE',
      isSortReverse: false
    };

    this.onSort = this.onSort.bind(this);
  }

  // sort trigger
  onSort(sortKey) {
    const isSortReverse =
      this.state.sortKey === sortKey && !this.state.isSortReverse;
    this.setState({ sortKey, isSortReverse });
  }

  render() {
    const { list, onDismiss } = this.props;

    const { sortKey, isSortReverse } = this.state;
    // const for sorted list with sort function
    const sortedList = SORTS[sortKey](list);
    const reverseSortedList = isSortReverse ? sortedList.reverse() : sortedList;

    return (
      <div className="table">
        <div className="table-header">
          <span>
            <Sort
              sortKey={'TITLE'}
              onSort={this.onSort}
              activeSortKey={sortKey}
            >
              Title <img src={sortArrows} alt="sort" />
            </Sort>
          </span>
          <span>
            <Sort
              sortKey={'AUTHOR'}
              onSort={this.onSort}
              activeSortKey={sortKey}
            >
              Author <img src={sortArrows} alt="sort" />
            </Sort>
          </span>
          <span>
            <Sort
              sortKey={'COMMENTS'}
              onSort={this.onSort}
              activeSortKey={sortKey}
            >
              Comments <img src={sortArrows} alt="sort" />
            </Sort>
          </span>
          <span>
            <Sort
              sortKey={'POINTS'}
              onSort={this.onSort}
              activeSortKey={sortKey}
            >
              Points <img src={sortArrows} alt="sort" />
            </Sort>
          </span>
          <span>Archive</span>
        </div>
        {reverseSortedList.map(item => (
          <div key={Math.random(item.objectID)} className="table-row">
            <span>
              <a className="post-title" href={item.url}>
                {item.title || 'N/A'}
              </a>
            </span>
            <span>{item.author || 'N/A'}</span>
            <span>{item.num_comments || 'N/A'}</span>
            <span>{item.points || 'N/A'}</span>
            <span>
              <Button
                onClick={() => onDismiss(item.objectID)}
                className="button-inline post-dismiss"
              >
                Dismiss
              </Button>
            </span>
          </div>
        ))}
      </div>
    );
  }
}

export default Table;
