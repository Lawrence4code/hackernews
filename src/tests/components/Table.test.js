import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Table from '../../components/Table';

// Adpater setup for enzyme

Enzyme.configure({ adapter: new Adapter() });

describe('List component', () => {
  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' }
    ]
  };

  test('render List without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
  });

  test('should List compoenent correctly', () => {
    const component = renderer.create(<Table {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should show two item in the list', () => {
    const wrapper = shallow(<Table {...props} />);
    expect(wrapper.find('.table-row').length).toBe(2);
  });
});
