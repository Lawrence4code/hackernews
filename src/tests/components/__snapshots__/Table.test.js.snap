// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should List compoenent correctly 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Title 
        <img
          alt="sort"
          src="select-arrows.svg"
        />
      </button>
    </span>
    <span>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Author 
        <img
          alt="sort"
          src="select-arrows.svg"
        />
      </button>
    </span>
    <span>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Comments 
        <img
          alt="sort"
          src="select-arrows.svg"
        />
      </button>
    </span>
    <span>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Points 
        <img
          alt="sort"
          src="select-arrows.svg"
        />
      </button>
    </span>
    <span>
      Archive
    </span>
  </div>
  <div
    className="table-row"
  >
    <span>
      <a
        className="post-title"
      >
        1
      </a>
    </span>
    <span>
      1
    </span>
    <span>
      1
    </span>
    <span>
      2
    </span>
    <span>
      <button
        className="button-inline post-dismiss"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span>
      <a
        className="post-title"
      >
        2
      </a>
    </span>
    <span>
      2
    </span>
    <span>
      1
    </span>
    <span>
      2
    </span>
    <span>
      <button
        className="button-inline post-dismiss"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
